### PR TITLE
crates/sound: read buffers from guest memory rather than copying them

### PR DESCRIPTION
**Pull request by MatiasVara**
This commit forces the reading of the buffers in the available ring only after a period has elapsed. This only applies for buffers that have arrived after the START control msg. This fix is required because the guest driver moves buffers from the used to the avail ring without knowing if the content has been updated by the user application. To fix this, after a request is enqueued in the available ring, the device waits a period before reading from guest memory.

### CODE REVIEW

File: crates/sound/src/audio_backends/pipewire.rs [hunk 1]
```diff
@@ -411,13 +430,20 @@ impl AudioBackend for PwBackend {
...
                             .get_mut(stream_id as usize)
                             .expect("Stream does not exist");
+
                         let Some(buffer) = streams.buffers.front_mut() else {
                             return;
                         };
 
+                        // only for requests that have arrived after START
+                        if buffer.arrival_req_time.elapsed() < streams.period && buffer.stream_state == PCMState::Start {
+                            // TODO: replace this with silence
```
Comment (dorindabassey) [inline crates/sound/src/audio_backends/pipewire.rs]: I don't understand why you intend to use silence here after Start state?

**Comment by dorindabassey** [comment]
Just one general comment to clean up clippy warnings from the PR

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: That comment is because I am not sure if it is OK to sleep in the threadloop. I was wondering if it is better to just fill a buffer with silence instead of sleeping.

File: crates/sound/src/stream.rs [hunk 2]
```diff
@@ -89,7 +94,7 @@ type Result<T> = std::result::Result<T, Error>;
 ///         |              |           |          |         |
 ///         |              |<----------|          |         |
 /// ```
-#[derive(Debug, Default, Copy, Clone)]
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
```
Comment (bilelmoussaoui) [inline crates/sound/src/stream.rs]: Also implement Eq

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: I am not sure about using silence as a preferred replacement either. because IIUC from the PR description, the idea is for the device to induce a delay right? which you already handled in pipewire.rs:L456 hence, I'll suggest to remove this conditional block of code. except I'm missing something?

Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: With that condition block, I induce a delay that is about a period. Otherwise, I do not how much delay I am adding. Another solution could be: 
1. to remove that conditional block
2. add a log if situation arrives
3. in the future, if we observe that the delay is not enough, we review the approach.

**Comment by MatiasVara** [comment]
I add @epilys to the discussion.

File: crates/sound/src/stream.rs [hunk 3]
```diff
@@ -189,6 +195,7 @@ impl Default for Stream {
             channels_max: 6,
             state: Default::default(),
             buffers: VecDeque::new(),
+            period: Duration::from_millis(0),
```
Comment (bilelmoussaoui) [inline crates/sound/src/stream.rs]: Why not just Duration::default?

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: By thinking again about this, we can remove that conditional block if we are sure that pw is calling process() in the right moment. Before, we were reading from guest's memory before process() was invoked. Now, we read from guest in the right moment which is when process() is invoked by pw.

Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: IMHO, i don't think this delay is needed

**Comment by stefano-garzarella** [comment]
> I add @epilys to the discussion.

Thanks!

@epilys this fix should be related to the "strange behavior" I mentioned in your v9

File: crates/sound/src/audio_backends/pipewire.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: I think you are right. I will remove it for the moment. Thanks.

**Review by dorindabassey — APPROVED** [review]

**Comment by stefano-garzarella** [comment]
@epilys @stsquad can you take a look?

This is a workaround for the potential spec violation of the current virtio-sound Linux driver.

File: crates/sound/src/audio_backends/pipewire.rs [hunk 4]
```diff
@@ -245,7 +245,7 @@ impl AudioBackend for PwBackend {
             let mut stream_hash = self.stream_hash.write().unwrap();
             let mut stream_listener = self.stream_listener.write().unwrap();
             self.thread_loop.lock();
-            let stream_params = self.stream_params.read().unwrap();
+            let stream_params = self.stream_params.write().unwrap();
```
Comment (stefano-garzarella) [inline crates/sound/src/audio_backends/pipewire.rs]: Why we need a `write` lock now?

File: crates/sound/src/audio_backends/pipewire.rs [hunk 5]
```diff
@@ -403,21 +403,22 @@ impl AudioBackend for PwBackend {
                         }
 
                         // to calculate as sizeof(int16_t) * NR_CHANNELS
-                        let frame_size = info.channels as u32 * size_of::<i16>() as u32;
+                        let frame_size = info.channels * size_of::<i16>() as u32;
```
Comment (stefano-garzarella) [inline crates/sound/src/audio_backends/pipewire.rs]: If this is not related, please move to another commit, since we may want to revert this commit, or keep this commit alone to work with the wrong Linux driver

File: crates/sound/src/audio_backends/pipewire.rs (hunk 4 above)
Comment (MatiasVara) [reply crates/sound/src/audio_backends/pipewire.rs]: We do not need a `write` lock. I'll change it back to a `read` lock.

File: crates/sound/src/stream.rs [hunk 6]
```diff
@@ -234,29 +235,41 @@ impl Default for PcmParams {
...
-            bytes,
             pos: 0,
+            data_descriptor,
             message,
         }
     }
+
+    pub fn consume(&self, buf: &mut [u8]) -> u32 {
+        let addr = self.data_descriptor.addr();
+        let offset : u64 = self.pos as u64;
```
Comment (bilelmoussaoui) [inline crates/sound/src/stream.rs]:
```suggestion
        let offset = self.pos as u64;
```

**Comment by stefano-garzarella** [comment]
The code LGTM, I just wonder if we should add a nice comment somewhere to explain why this behavior is happening (like in the commit description), just so we don't change it in the future. (assuming the driver is behaving correctly, but for now we need this behavior in the device until we decide what to do with the driver)

**Comment by MatiasVara** [comment]
> The code LGTM, I just wonder if we should add a nice comment somewhere to explain why this behavior is happening (like in the commit description), just so we don't change it in the future. (assuming the driver is behaving correctly, but for now we need this behavior in the device until we decide what to do with the driver)

I agree. I could add a comment when processing the TX queue to clarify why we store a sort of pointer to the data descriptor instead of directly copying its content. WHDYT?

**Comment by stefano-garzarella** [comment]
> > The code LGTM, I just wonder if we should add a nice comment somewhere to explain why this behavior is happening (like in the commit description), just so we don't change it in the future. (assuming the driver is behaving correctly, but for now we need this behavior in the device until we decide what to do with the driver)
> 
> I agree. I could add a comment when processing the TX queue to clarify why we store a sort of pointer to the data descriptor instead of directly copying its content. WHDYT?

Ack ;-)

File: crates/sound/src/audio_backends/alsa.rs [hunk 7]
```diff
@@ -191,15 +193,17 @@ fn write_samples_io(
             let Some(buffer) = stream.buffers.front_mut() else {
                 return 0;
             };
-            let mut iter = buffer.bytes[buffer.pos..].iter().cloned();
+            let mut buf2 = vec![0; buffer.data_descriptor.len() as usize];
```
Comment (stefano-garzarella) [inline crates/sound/src/audio_backends/alsa.rs]: `buf2` is a bit cryptic IMHO, what about `data`?

File: crates/sound/src/device.rs [hunk 8]
```diff
@@ -469,14 +469,16 @@ impl VhostUserSoundThread {
...
-                        buf.truncate(bytes_read);
-
-                        buffers.push(Buffer::new(buf, Arc::clone(&message)));
+                        /*
+                        Rather than copying the content of a descriptor, buffer keeps a pointer to it.
+                        When we copy just after the request is enqueued, the guest's userspace may or
+                        may not have updated the buffer contents. Guest driver simply moves buffers
+                        from the used ring to the available ring without knowing whether the content
+                        has been updated. The device only reads the buffer from guest memory when the
+                        audio engine requires it. The delay added by this allows the buffer to be
```
Comment (stefano-garzarella) [inline crates/sound/src/device.rs]: More than a delay, does this create a kind of synchronization between the producer (driver) and consumer (audio backend) based on the negotiated period?

File: crates/sound/src/stream.rs [hunk 9]
```diff
@@ -89,7 +89,7 @@ type Result<T> = std::result::Result<T, Error>;
 ///         |              |           |          |         |
 ///         |              |<----------|          |         |
 /// ```
-#[derive(Debug, Default, Copy, Clone)]
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
```
Comment (stefano-garzarella) [inline crates/sound/src/stream.rs]: why we need this change?

File: crates/sound/src/stream.rs [hunk 10]
```diff
@@ -234,29 +234,46 @@ impl Default for PcmParams {
...
+            .message
+            .desc_chain
+            .memory()
+            .read(
+                buf,
+                addr.checked_add(offset)
+                    .expect("invalid guest memory address"),
+            )
+            .map_err(|_| StreamError::DescriptorReadFailed)
+            .unwrap();
```
Comment (stefano-garzarella) [inline crates/sound/src/stream.rs]: Should be better to propagate the error?

File: crates/sound/src/stream.rs [hunk 11]
```diff
@@ -4,9 +4,9 @@
 use std::{collections::VecDeque, sync::Arc};
 
 use thiserror::Error as ThisError;
-use vm_memory::{Le32, Le64};
+use vm_memory::{Address, Bytes, Le32, Le64};
 
-use crate::{virtio_sound::*, IOMessage, SUPPORTED_FORMATS, SUPPORTED_RATES};
+use crate::{virtio_sound::*, Error as StreamError, IOMessage, SUPPORTED_FORMATS, SUPPORTED_RATES};
```
Comment (stefano-garzarella) [inline crates/sound/src/stream.rs]: `StreamError` is a bit confusing here since this is like our generic error enum at the crate level.
I'd suggest something like `SoundError`.

File: crates/sound/src/stream.rs (hunk 9 above)
Comment (MatiasVara) [reply crates/sound/src/stream.rs]: I added based on a comment here https://github.com/virtio-sound/vhost-device/pull/32#discussion_r1325627215. However, I think that comment is now outdated. I will drop this change.

Comment (stefano-garzarella) [reply crates/sound/src/stream.rs]: BTW I agree with that comment, my question was more about why we need both. If not needed for this fix, I agree on remove the entire change.

File: crates/sound/src/device.rs (hunk 8 above)
Comment (MatiasVara) [reply crates/sound/src/device.rs]: Would it work to replace last two sentences by ``The device only reads the buffer from guest memory when the audio engine requires it, which is after a period thus ensuring that the buffer is up-to-date.``?

Comment (stefano-garzarella) [reply crates/sound/src/device.rs]: yep, looks better ;-)

Comment (stsquad) [reply crates/sound/src/device.rs]: Is anyone looking at fixing the driver itself. This seems like a workround for the driver violating the spec.

Comment (stefano-garzarella) [reply crates/sound/src/device.rs]: Yep, Matias reported it upstream and they are discussing next steps for the driver and specs: https://lists.oasis-open.org/archives/virtio-comment/202309/msg00175.html

Comment (MatiasVara) [reply crates/sound/src/device.rs]: This may be a workaround, but I'm not sure. Even if the driver behaves correctly, the device can't consume requests faster than (requests per period) for a given stream. Instead of copying the buffer content, this commit keeps a "pointer" instead, which could also be considered an optimization over keeping a copy of it.

Comment (stsquad) [reply crates/sound/src/device.rs]: I guess the buffer refs ensure we never return a buffer until it has finished being consumed?

Comment (MatiasVara) [reply crates/sound/src/device.rs]: This commit did not change the way we return buffers. This still happens after a whole period_bytes has been consumed and a MessageIO is dropped. Consume here means have read from guest memory.

**Review by stefano-garzarella — APPROVED** [review]
